src/File/RenameUpload.php

```diff
@@ -171,32 +173,49 @@ public function filter($value)
             }
 
             $isFileUpload = true;
-            $uploadData = $value;
             $sourceFile = $value['tmp_name'];
+            $clientFilename = $value['name'];
+        } elseif ($value instanceof UploadedFileInterface) {
+            $isFileUpload = true;
+            $sourceFile = $value->getStream()->getMetadata('uri');
+            $clientFilename = $value->getClientFilename();
         } else {
-            $uploadData = [
-                'tmp_name' => $value,
-                'name'     => $value,
-            ];
+            $clientFilename = $value;
             $sourceFile = $value;
         }
 
         if (isset($this->alreadyFiltered[$sourceFile])) {
             return $this->alreadyFiltered[$sourceFile];
         }
 
-        $targetFile = $this->getFinalTarget($uploadData);
-        if (! file_exists($sourceFile) || $sourceFile == $targetFile) {
+        $targetFile = $this->getFinalTarget($sourceFile, $clientFilename);
+        if ($sourceFile === $targetFile || ! file_exists($sourceFile)) {
             return $value;
         }
 
         $this->checkFileExists($targetFile);
-        $this->moveUploadedFile($sourceFile, $targetFile);
+        if ($value instanceof UploadedFileInterface) {
+            $value->moveTo($targetFile);
+        } else {
+            $this->moveUploadedFile($sourceFile, $targetFile);
+        }
 
         $return = $targetFile;
         if ($isFileUpload) {
-            $return = $uploadData;
-            $return['tmp_name'] = $targetFile;
+            if ($value instanceof UploadedFileInterface) {
+                $return = new UploadedFile(
+                    $targetFile,
+                    filesize($targetFile),
+                    UPLOAD_ERR_OK,
+                    $value->getClientFilename(),
+                    $value->getClientMediaType()
+                );
+            } else {
+                $return = [
+                    'tmp_name' => $clientFilename,
+                    'name'     => $targetFile,
+                ];
+            }
```

Comment:
I'd like to refactor this entire method. The logic is ridiculously convoluted. We should extract a method per `$value` type, and call the appropriate method based on type to perform the work.

I'll start work on that immediately.